File: app/src/main/java/com/canchapp_kotlin/ui/complex/ComplexViewModel.kt
package com.canchapp_kotlin.ui.complex

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.canchapp_kotlin.data.network.ComplexListResponse
import com.canchapp_kotlin.data.repository.ComplexRepository
import com.canchapp_kotlin.util.Resource
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ComplexViewModel : ViewModel() {

    private val repository = ComplexRepository()

    private val _complexesState = MutableStateFlow<Resource<ComplexListResponse>?>(null)
    val complexesState = _complexesState.asStateFlow()

    init {
        loadComplexes()
    }

    fun loadComplexes(page: Int = 1) {
        viewModelScope.launch {
            _complexesState.value = Resource.Loading
            _complexesState.value = repository.getComplexes(page)
        }
    }
}

File: ios/CanchApp/ViewModels/ComplexViewModel.swift
import Foundation

/// Equivalent of Kotlin's ComplexViewModel.
@MainActor
final class ComplexViewModel: ObservableObject {

    @Published var complexesState: Resource<ComplexListResponse>? = nil

    private var userLat: Double? = nil
    private var userLon: Double? = nil
    private let repository = ComplexRepository()

    init() {
        loadComplexes()
    }

    func updateLocation(lat: Double, lon: Double) {
        userLat = lat
        userLon = lon
        loadComplexes()
    }

    func loadComplexes(page: Int = 1) {
        complexesState = .loading
        Task {
            complexesState = await repository.getComplexes(page: page, lat: userLat, lon: userLon)
        }
    }
}
